Failure to build on manylinux_2_28

Attempting this:

```shellsession
$ docker pull quay.io/pypa/manylinux_2_28_aarch64:2025.05.16-1
...

$ docker run -it quay.io/pypa/manylinux_2_28_aarch64:2025.05.16-1 bash
# yum update && yum install -y libffi-devel
...
# python3.13t -m pip install git+https://github.com/Quansight-Labs/cffi.git
```

Output (collapsed):

<details>

```
Collecting git+https://github.com/Quansight-Labs/cffi.git
  Cloning https://github.com/Quansight-Labs/cffi.git to /tmp/pip-req-build-zu1ld_xr
  Running command git clone --filter=blob:none --quiet https://github.com/Quansight-Labs/cffi.git /tmp/pip-req-build-zu1ld_xr
  Resolved https://github.com/Quansight-Labs/cffi.git to commit c06e9e145e3b0eda373538e89ff85db99a33d62c
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting pycparser (from cffi==1.18.0.dev0)
  Using cached pycparser-2.22-py3-none-any.whl.metadata (943 bytes)
Using cached pycparser-2.22-py3-none-any.whl (117 kB)
Building wheels for collected packages: cffi
  Building wheel for cffi (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for cffi (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [97 lines of output]
      /tmp/pip-build-env-a9ltoxln/overlay/lib/python3.13t/site-packages/setuptools/config/_apply_pyprojecttoml.py:61: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!

              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:

              License :: OSI Approved :: MIT License

              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************

      !!
        dist._finalize_license_expression()
      /tmp/pip-build-env-a9ltoxln/overlay/lib/python3.13t/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!

              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:

              License :: OSI Approved :: MIT License

              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************

      !!
        self._finalize_license_expression()
      running bdist_wheel
      running build
      running build_py
      creating build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/__init__.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/_imp_emulation.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/_shimmed_dist_utils.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/api.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/backend_ctypes.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/cffi_opcode.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/commontypes.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/cparser.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/error.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/ffiplatform.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/lock.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/model.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/pkgconfig.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/recompiler.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/setuptools_ext.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/vengine_cpy.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/vengine_gen.py -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/verifier.py -> build/lib.linux-aarch64-cpython-313t/cffi
      running egg_info
      writing src/cffi.egg-info/PKG-INFO
      writing dependency_links to src/cffi.egg-info/dependency_links.txt
      writing entry points to src/cffi.egg-info/entry_points.txt
      writing requirements to src/cffi.egg-info/requires.txt
      writing top-level names to src/cffi.egg-info/top_level.txt
      reading manifest file 'src/cffi.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      adding license file 'LICENSE'
      adding license file 'AUTHORS'
      writing manifest file 'src/cffi.egg-info/SOURCES.txt'
      copying src/cffi/_cffi_errors.h -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/_cffi_include.h -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/_embedding.h -> build/lib.linux-aarch64-cpython-313t/cffi
      copying src/cffi/parse_c_type.h -> build/lib.linux-aarch64-cpython-313t/cffi
      running build_ext
      building '_cffi_backend' extension
      creating build/temp.linux-aarch64-cpython-313t/src/c
      gcc -pthread -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall -fPIC -DFFI_BUILDING=1 -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/opt/_internal/cpython-3.13.3-nogil/include/python3.13t -c src/c/_cffi_backend.c -o build/temp.linux-aarch64-cpython-313t/src/c/_cffi_backend.o
      src/c/_cffi_backend.c: In function ‘b_do_dlopen’:
      src/c/_cffi_backend.c:4625:22: warning: ‘Py_FileSystemDefaultEncoding’ is deprecated [-Wdeprecated-declarations]
       4625 |                      Py_FileSystemDefaultEncoding, &filename_or_null, &flags))
            |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      In file included from /opt/_internal/cpython-3.13.3-nogil/include/python3.13t/Python.h:99,
                       from src/c/_cffi_backend.c:2:
      /opt/_internal/cpython-3.13.3-nogil/include/python3.13t/fileobject.h:22:46: note: declared here
         22 | Py_DEPRECATED(3.12) PyAPI_DATA(const char *) Py_FileSystemDefaultEncoding;
            |                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      src/c/_cffi_backend.c: In function ‘get_unique_type’:
      src/c/_cffi_backend.c:4756:37: error: passing argument 1 of ‘PyBytes_FromStringAndSize’ from incompatible pointer type [-Wincompatible-pointer-types]
       4756 |     key = PyBytes_FromStringAndSize(unique_key, keylength * sizeof(void *));
            |                                     ^~~~~~~~~~
            |                                     |
            |                                     const void **
      In file included from /opt/_internal/cpython-3.13.3-nogil/include/python3.13t/Python.h:78:
      /opt/_internal/cpython-3.13.3-nogil/include/python3.13t/bytesobject.h:31:50: note: expected ‘const char *’ but argument is of type ‘const void **’
         31 | PyAPI_FUNC(PyObject *) PyBytes_FromStringAndSize(const char *, Py_ssize_t);
            |                                                  ^~~~~~~~~~~~
      In file included from src/c/cffi1_module.c:3,
                       from src/c/_cffi_backend.c:8126:
      src/c/realize_c_type.c: In function ‘init_global_types_dict’:
      src/c/realize_c_type.c:53:14: error: assignment to ‘CTypeDescrObject *’ {aka ‘struct _ctypedescr *’} from incompatible pointer type ‘PyObject *’ {aka ‘struct _object *’} [-Wincompatible-pointer-types]
         53 |     g_ct_int = get_primitive_type(_CFFI_PRIM_INT);         // 'int'
            |              ^
      src/c/realize_c_type.c: In function ‘_realize_c_struct_or_union’:
      src/c/realize_c_type.c:373:16: error: returning ‘CTypeDescrObject *’ {aka ‘struct _ctypedescr *’} from a function with incompatible return type ‘PyObject *’ {aka ‘struct _object *’} [-Wincompatible-pointer-types]
        373 |         return g_file_struct;
            |                ^~~~~~~~~~~~~
      error: command '/opt/rh/gcc-toolset-14/root/usr/bin/gcc' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for cffi
Failed to build cffi
ERROR: Failed to build installable wheels for some pyproject.toml based projects (cffi)
```

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -287,7 +287,7 @@ jobs:
           cd .. && \
           rm -rf libffi-3.4.6
         CIBW_ENVIRONMENT_PASS_LINUX: CFLAGS  # ensure that the build container can see our overridden build config
-        CIBW_MANYLINUX_X86_64_IMAGE: ${{ matrix.manylinux_img || '' }}
+        CIBW_MANYLINUX_X86_64_IMAGE: ${{ matrix.manylinux_img || 'manylinux_2_28' }}
         CIBW_MANYLINUX_I686_IMAGE: ${{ matrix.manylinux_img || '' }}
         CIBW_MANYLINUX_AARCH64_IMAGE: ${{ matrix.manylinux_img || '' }}
         CIBW_MANYLINUX_PPC64LE_IMAGE: ${{ matrix.manylinux_img || '' }}

--- a/src/c/_cffi_backend.c
+++ b/src/c/_cffi_backend.c
@@ -4753,7 +4753,7 @@ static PyObject *get_unique_type(CTypeDescrObject *x,
     */
     PyObject *key, *y;
 
-    key = PyBytes_FromStringAndSize(unique_key, keylength * sizeof(void *));
+    key = PyBytes_FromStringAndSize((const char *)unique_key, keylength * sizeof(void *));
     if (key == NULL) {
         Py_DECREF(x);
         return NULL;
@@ -4773,7 +4773,7 @@ static void remove_dead_unique_reference(PyObject *unique_key)
 {
     PyObject *wr;
     PyObject *tmp = NULL;
-    int err;
+    int err = 0;
 
     LOCK_UNIQUE_CACHE();
     /* We need to check that it's not already been replaced by a live weakref

--- a/src/c/realize_c_type.c
+++ b/src/c/realize_c_type.c
@@ -50,7 +50,7 @@ static int init_global_types_dict(PyObject *ffi_type_dict)
         return -1;
     g_ct_chararray = (CTypeDescrObject *)ct2;
 
-    g_ct_int = get_primitive_type(_CFFI_PRIM_INT);         // 'int'
+    g_ct_int = (CTypeDescrObject *)get_primitive_type(_CFFI_PRIM_INT);    // 'int'
     if (g_ct_int == NULL)
         return -1;
 
@@ -370,7 +370,7 @@ _realize_c_struct_or_union(builder_c_t *builder, int sindex)
     if (sindex == _CFFI__IO_FILE_STRUCT) {
         /* returns a single global cached opaque type */
         Py_INCREF(g_file_struct);
-        return g_file_struct;
+        return (PyObject *)g_file_struct;
     }
 
     s = &builder->ctx.struct_unions[sindex];

